src/engine/SPARQLProtocol.h

```diff
@@ -205,5 +105,190 @@ class SPARQLProtocol {
+  }
+
+  static ad_utility::url_parser::ParsedRequest parseUpdatePOST(
+      const ad_utility::httpUtils::HttpRequest auto& request) {
+    return parseSPARQLPOST<ad_utility::url_parser::sparqlOperation::Update>(
+        request, contentTypeSparqlUpdate);
+  }
+
+  static ad_utility::url_parser::ParsedRequest parsePOST(
+      const ad_utility::httpUtils::HttpRequest auto& request) {
+    // For a POST request, the content type *must* be either
+    // "application/x-www-form-urlencoded" (1), "application/sparql-query"
+    // (2) or "application/sparql-update" (3).
+    //
+    // (1) Section 2.1.2: The body of the POST request contains *all*
+    // parameters (including the query or update) in an encoded form (just
+    // like in the part of a GET request after the "?").
+    //
+    // (2) Section 2.1.3: The body of the POST request contains *only* the
+    // unencoded SPARQL query. There may be additional HTTP query parameters.
+    //
+    // (3) Section 2.2.2: The body of the POST request contains *only* the
+    // unencoded SPARQL update. There may be additional HTTP query parameters.
+    //
+    // Reference: https://www.w3.org/TR/2013/REC-sparql11-protocol-20130321
+    std::string_view contentType =
+        request.base()[boost::beast::http::field::content_type];
+    LOG(DEBUG) << "Content-type: \"" << contentType << "\"" << std::endl;
+
+    // Note: For simplicity we only check via `starts_with`. This ignores
+    // additional parameters like `application/sparql-query;charset=utf8`. We
+    // currently always expect UTF-8.
+    // TODO<joka921> Implement more complete parsing that allows the checking
+    // of these parameters.
+    if (contentType.starts_with(contentTypeUrlEncoded)) {
+      return parseUrlencodedPOST(request);
+    }
+    if (contentType.starts_with(contentTypeSparqlQuery)) {
+      return parseQueryPOST(request);
+    }
+    if (contentType.starts_with(contentTypeSparqlUpdate)) {
+      return parseUpdatePOST(request);
+    }
+    // Checking if the content type is supported by the Graph Store
+    // HTTP Protocol implementation is done later.
+    auto parsedRequestBuilder = ParsedRequestBuilder(request);
+    if (parsedRequestBuilder.isGraphStoreOperation()) {
+      parsedRequestBuilder.extractGraphStoreOperation();
+      parsedRequestBuilder.extractAccessToken(request);
+      return std::move(parsedRequestBuilder).build();
+    }
+
+    throw std::runtime_error(absl::StrCat(
+        "POST request with content type \"", contentType,
+        "\" not supported (must be \"", contentTypeUrlEncoded, "\", \"",
+        contentTypeSparqlQuery, "\" or \"", contentTypeSparqlUpdate, "\")"));
+  }
+
+ public:
+  /// Parse the path and URL parameters from the given request. Supports both
+  /// GET and POST request according to the SPARQL 1.1 standard.
+  static ad_utility::url_parser::ParsedRequest parseHttpRequest(
+      const ad_utility::httpUtils::HttpRequest auto& request) {
```

Comment:
In general: 
I think you assume that the templated request has a `string body` and some map of parameters, can't you extract them initially, and then have ordinary non-templated code, that can live in the `.cpp`file?

Reply:
The model is a string body, a `ParsedUr` (string path and http query parameter map) and an optional header authorization field.

We gain the type of the request body with the templating of the `HttpRequest`. But afaik it is assumed to be string anyways in most (all?) places. The templating on the other hand causes us quite some effort. Was there a reason why this way was chosen?